system-addon/lib/Screenshots.jsm

```diff
@@ -63,20 +65,32 @@ this.Screenshots = {
     return screenshot;
   },
 
+  /**
+  * Update image cache so that it will use the custom screenshot when
+  * looking up the site url.
+  *
+  * @param newScreenshotURL {string} URL pointing to the custom screenshot
+  * @param url {string} Site url used to lookup the screenshot.
+  */
+  async replaceScreenshot(newScreenshotURL, url) {
+    // .copy() will overwrite any existing thumbnail associated with `url`
+    await PageThumbsStorage.copy(newScreenshotURL, url);
```

Comment:
This doesn't seem right. There are other consumers of `PageThumbs`, so having it show some custom image that is *not* the page's thumbnail will probably lead to bugs. Highlights doesn't replace a url's thumbnail with the `preview_image_url` by requesting the thumbnail for that image instead of the page. Maybe something like that could be used for these custom preview images for top sites?

Reply:
Changed this. Now a preview request will trigger a `getScreenshotForURL` and we attach that in the Reducer to the Topsite as `screenshotPreview`.
If the user saves the form (it pins the site) we save the URL in the pinned pref as customScreenshotURL and upon the feed refresh we trigger `maybeCacheScreenshot` which adds `customScreenshot` to the LinksCache object that represents that Topsite. The `customScreenshot` is also migrated between links so we need to clear the value if the user updated the customScreenshotURL for the website.